src/marqo/core/document/add_documents_handler.py

```diff
@@ -0,0 +1,562 @@
+        # TODO we ignore the location for now, and will add it if needed in the future
+        self.responses: List[Tuple[int, MarqoAddDocumentsItem]] = []
+        self.errors = False
+        self.marqo_docs: Dict[str, Dict[str, Any]] = dict()
+        self.marqo_doc_loc_map: Dict[str, int] = dict()
+        self.visited_doc_ids: Set[str] = set()
+
+    def visited(self, doc_id: str) -> bool:
+        return doc_id in self.visited_doc_ids
+
+    def collect_marqo_doc(self, loc: int, marqo_doc: Dict[str, Any]):
+        doc_id = marqo_doc[MARQO_DOC_ID]
+        self.marqo_docs[doc_id] = marqo_doc
+        self.marqo_doc_loc_map[doc_id] = loc
+        if marqo_doc[ORIGINAL_ID] is not None:
+            self.visited_doc_ids.add(marqo_doc[ORIGINAL_ID])
+
+    def collect_error_response(self, doc_id: Optional[str], error: AddDocumentsError, loc: Optional[int] = None):
+        # log errors in one place, log in warn level for each individual doc error
+        # TODO it might be too verbose, but check if we need exc_info=(type(error), error, error.__traceback__)
+        logger.warning(f'Encountered error when adding doc {doc_id}: {str(error)}')
+
+        if isinstance(error, DuplicateDocumentError):
+            # This is the current logic, docs with same id supersedes previous ones defined in the batch
+            # TODO change the logic when we need to report duplicates as error in the response
+            return
+
+        if not loc and doc_id and doc_id in self.marqo_doc_loc_map:
+            loc = self.marqo_doc_loc_map[doc_id]
+
+        if doc_id in self.marqo_docs:
+            doc_id = self.marqo_docs.pop(doc_id)[ORIGINAL_ID]
+
+        # Even if the last document is invalid, we should not use previous ones?
+        if doc_id:
+            self.visited_doc_ids.add(doc_id)
+
+        self.responses.append((loc, MarqoAddDocumentsItem(
+            id=doc_id if doc_id is not None else '',
+            error=error.error_message,
+            message=error.error_message,
+            status=error.status_code,
+            code=error.error_code
+        )))
+
+        self.errors = True
+
+    def collect_successful_response(self, doc_id: Optional[str]):
+        loc = self.marqo_doc_loc_map.get(doc_id, None)
+
+        self.responses.append((loc, MarqoAddDocumentsItem(
+            id=doc_id if doc_id is not None else '',
+            status=200,
+        )))
+
+    def to_add_doc_responses(self, index_name: str) -> MarqoAddDocumentsResponse:
+        processing_time = (timer() - self.start_time) * 1000
+        # since we reversed the doc list to skip duplicate docs, we now need to reverse the response
+        sorted_responses = [response for _, response in sorted(self.responses, key=lambda r: r[0] or 0, reverse=True)]
+        return MarqoAddDocumentsResponse(errors=self.errors, index_name=index_name, items=sorted_responses,
+                                         processingTimeMs=processing_time)
+
+
+class AddDocumentsHandler(ABC):
```

Comment:
Ideally we want one core class Document that provides all functionality. I find the abstractions a bit confusing at this point as `Document` handles partial updates. Perhaps we could have these handlers for add docs, update etc and have `Document` as a facade that everyone else calls?

Reply:
I think adding documents is quite complex, it makes sense to have its own class.  I like your idea to make Document as a facade to call this handler. (I'll move the add_documents method in tensor_search in Documents at the end of refactoring) For other logic like deletion and partial update, I'm not sure if it worth having their own handler. If the logic is generic, we won't need a class hierarchy to capture index type specific logic, then it makes sense to keep it in Documents class.